use vibe.d's cryptorand to generate state parameter

MartinNowak opened the pull request:
- secrets based on timestamps and PIDs are not secure
- also add unittest

MartinNowak committed:
use vibe.d's cryptorand to generate state parameter

- secrets based on timestamps and PIDs are not secure
- also add unittest

thaven commented on source/oauth/settings.d:
```diff
@@ -387,22 +432,8 @@ class OAuthSettings
-            data[16 .. 20] = crc32Of(this.scopes);
-            return sha256Of(data[]);
-        }
+        ubyte[16] key;
```
By this change, LoginData.key loses its dependency on `timestamp` and `scopes`, which had a purpose.

thaven reviewed (changes requested):
I understand what you say about secrets based on timestamps, but I think you miss the point here. Have a look at [line 330 of settings.d](https://github.com/MartinNowak/oauth/blob/3b77eb3084f7710e15d16ac344a0e49b0ebd667f/source/oauth/settings.d#L330). Though unmodified, this totally changes meaning. Where it used to be a check if the `LoginData` stored in the session is the same as the one used to construct the authUri, it is now reduced to just a check if the state parameter matches the random key. So any tampering with the scope parameter, for example, will now go undetected. The randomId in the original implementation just adds a little randomness to reduce the chance of state key not being unique.

MartinNowak committed:
use hmac over timestamp, scopes, and random secret as state

- ensures integrity of those values though they are stored in the session
- also use constant time compare for the state parameter (>=2.075)

MartinNowak commented:
I thought the state is mostly a CSRF token, no?

> I understand what you say about secrets based on timestamps, but I think you miss the point here. Have a look at line 330 of settings.d. Though unmodified, this totally changes meaning. Where it used to be a check if the LoginData stored in the session is the same as the one used to construct the authUri, it is now reduced to just a check if the state parameter matches the random key. So any tampering with the scope parameter, for example, will now go undetected. The randomId in the original implementation just adds a little randomness to reduce the chance of state key not being unique.

Well I've now used a proper HMAC to sign the data with a random secret.
Though I don't really see why you'd want to verify the LoginData, as that comes from your session store and the random data is now really unique.

This also uses [secureEqual](https://dlang.org/library/std/digest/secure_equal.html) now, so that the state cannot be easily determined by timing attacks.
https://dlang.org/changelog/2.075.0.html#std-digest-digest-secureCompare

thaven commented:
> Though I don't really see why you'd want to verify the LoginData, as that comes from your session store

Okay, so just while I started typing my motivation for that, I realize it doesn't even work as intended in the original implementation.

I was trying to solve two little problems at once here, but now I realize I only solved one.

Both problems stem from the fact that the OAuth spec does not require the authorization server to return the scopes granted, if they match the ones requested (spec bug, IMO). So, it is necessary to know the scopes requested, based on the state parameter. That's the reason for LoginData to exist in the first place. Though, it got used for login attempt expiration as well.

First problem would occur if two authUrls are generated with different scopes. The second attempt would overwrite the first LoginData, and a subsequent login using the authUrl generated first may trick the client to assume it got granted the scopes from the second attempt.

Second problem is that, any authorization server that does omit the scopes from the response, also allows one to trick the client into believing it got granted some scopes that were not really granted, just by modifying the redirect url.

Such wrong assumptions about scopes granted may cause errors when actually accessing a resource using the obtained token. Such would be the kind of error that could not happen in theory (we checked the scope, right?), so the most likely not to be handled properly.

So, I solved the first one by signing LoginData and using the signature as a state parameter. If the scopes get changed in the redirect url, the state parameter is not a valid signature anymore. Unfortunately, we never really get to check that, so I did not solve the second problem at all. I even think that one is impossible to solve.

Conclusion: you're right, the signature doesn't add much. A random key already solves the first problem, and the second problem wasn't solved by the signature anyway.

thaven commented:
But using the HMAC signature we can get rid of the pre-login session usage...

thaven commented:
@MartinNowak can you resolve the conflict, please? I'm not sure where the -ddox-tool line should go.

MartinNowak commented:
Done @thaven, it's a top-level parameter.
